### ISSUE TITLE
Math functions like acos() conflict with their declaration in math.h

### ISSUE THREAD

**Issue by k15tfu** [dotnet/coreclr#20786]
In `src/pal/inc/pal.h` we define acos() the different way it is defined on my computer (not Ubuntu 14.04).  The difference is in the linkage type and in the exception specification.

So if pal.h is somehow included (for example, [via assert.h](https://github.com/dotnet/coreclr/issues/20784)) together with math.h I get compile errors:
```
In file included from /usr/include/math.h:290:
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:143:13: error: conflicting types for 'sqrt'
__MATHCALL (sqrt,, (_Mdouble_ __x));
            ^
/home/user/p/coreclr/src/pal/inc/pal.h:4411:26: note: previous declaration is here
PALIMPORT double __cdecl sqrt(double);
```

The simplest way to fix it is to ensure math.h is included before these declarations in pal.h. In this case, there will be no warnings. So we have to just include math.h there and everything will be fine: https://github.com/dotnet/coreclr/blob/master/src/pal/inc/pal.h#L46 .

Example: https://godbolt.org/z/imzEtp

P.S. I can prepare a pull request.

**Comment by tannergooding** [comment]
How/where are you getting `math.h` included such that these errors will appear?

The way this is supposed to work is that the real `<math.h>` is only included by https://github.com/dotnet/coreclr/blob/master/src/pal/src/cruntime/math.cpp (effectively) and the rest of the JIT/VM includes https://github.com/dotnet/coreclr/blob/master/src/pal/inc/rt/cpp/math.h. This allows us to redefine certain math functions, as necessary, so that we can ensure they return the IEEE compliant values where required.

For example:
* We define macros that go from `acos` to `PAL_acos` here: https://github.com/dotnet/coreclr/blob/master/src/pal/inc/pal.h#L4155.
* We then declare the function signature for `PAL_acos` here (relying on the macro to change it to the correct name): https://github.com/dotnet/coreclr/blob/master/src/pal/inc/pal.h#L4412
* We then provide an implementation for `PAL_acos` here: https://github.com/dotnet/coreclr/blob/master/src/pal/src/cruntime/math.cpp#L142
  * We undefine the macro introduced in pal.h here: https://github.com/dotnet/coreclr/blob/master/src/pal/src/include/pal/palinternal.h#L449
  * We use cmake to check for special values at compile time so we can adjust the function as needed: https://github.com/dotnet/coreclr/blob/master/src/pal/src/configure.cmake#L690
 
We do something similar for functions that are defined via macros themselves (such as `isnan`), but generally introduce a new method prefixed with `_` that we use throughout the rest of the codebase instead.

**Comment by tannergooding** [comment]
It sounds like something is still being misconfigured then. Either `_Mdouble_` isn't being set to the right type, `__MATHCALL` is using a different calling convention than expected, or something else in the actual signature is non-compatible. All of these could lead to unexpected breaks in behavior/functionality.

**Comment by k15tfu** [comment]
@tannergooding Well, what's the next?  Should I prepare the proposed pull request with math.h?

**Comment by tannergooding** [comment]
@jkotas, who is the right owner to tag for the PAL layer? The suggested fix (including `math.h` with the rest of the includes under the `PAL_STDCPP_COMPAT` define (https://github.com/dotnet/coreclr/blob/master/src/pal/inc/pal.h#L46) seems like a reasonable fix to me...

**Comment by tannergooding** [comment]
Ah, I just saw @janvorli's response here: https://github.com/dotnet/coreclr/issues/20766#issuecomment-439049513. The same thing might apply to this issue as well

**Comment by janvorli** [comment]
@tannergooding I am the PAL owner.

**Comment by janvorli** [comment]
@k15tfu I think that rather than including the math.h and keeping the definitions, it would actually be better if we just moved all the math functions declarations under `#ifndef PAL_STDCPP_COMPAT`. There is no point in keeping the declarations in that case.

**Comment by k15tfu** [comment]
@janvorli I can suggest only one reason to do like that:  It can help us to find conflicts between standard declarations and our "stubs": https://godbolt.org/z/CL8PHd .  But maybe this is a bit far-fetched =)

**Comment by k15tfu** [comment]
@janvorli One more thing: The same errors will appear for `rand`/`srand`/`exit` and many other functions if we remove `#include <stdlib.h>` from https://github.com/dotnet/coreclr/blob/master/src/pal/inc/pal.h#L42 .
```
In file included from /opt/clang+llvm-6.0.0-x86_64-linux-gnu-ubuntu-16.04/bin/../include/c++/v1/cstdlib:86:
In file included from /opt/clang+llvm-6.0.0-x86_64-linux-gnu-ubuntu-16.04/bin/../include/c++/v1/stdlib.h:94:
/usr/include/stdlib.h:592:12: error: conflicting types for 'atexit'
extern int atexit (void (*__func) (void)) __THROW __nonnull ((1));
           ^
/home/user/p/coreclr/src/pal/inc/pal.h:4477:13: note: previous declaration is here
int __cdecl atexit(void (__cdecl *function)(void));
            ^
```

I think the reason is because all C functions from standard library are declared with __THROW (i.e. noexcept).  So if you think about moving the math functions under `#ifndef PAL_STDCPP_COMPAT` (because they conflict with standard headers), we probably should also move all other declarations (like `malloc`/`rand`/`exit`) into this #ifdef.

Or, we can just include an appropriate header for these conflicting declarations (`math.h` in this case) as it's done for all others prototypes.

**Comment by k15tfu** [comment]
@janvorli I have prepared the PR as you suggested: https://github.com/dotnet/coreclr/pull/21170
